firebase-firestore/src/main/java/com/google/firebase/firestore/DocumentReference.java

```diff
@@ -455,7 +459,39 @@ public ListenerRegistration addSnapshotListener(
     checkNotNull(metadataChanges, "Provided MetadataChanges value must not be null.");
     checkNotNull(listener, "Provided EventListener must not be null.");
     return addSnapshotListenerInternal(
-        Executors.DEFAULT_CALLBACK_EXECUTOR, internalOptions(metadataChanges), activity, listener);
+        Executors.DEFAULT_CALLBACK_EXECUTOR,
+        internalOptions(metadataChanges),
+        activity,
+        null,
+        listener);
+  }
+
+  /**
+   * Starts listening to the document referenced by this {@code DocumentReference} with the given
+   * options using an LifecycleOwner-scoped listener.
+   *
+   * <p>The listener will be automatically removed during {@link LifecycleOwner's ON_STOP Event'}.
+   *
+   * @param lifecycleOwner The LifecycleOwner to scope the listener to.
+   * @param metadataChanges Indicates whether metadata-only changes (i.e. only {@code
+   *     DocumentSnapshot.getMetadata()} changed) should trigger snapshot events.
+   * @param listener The event listener that will be called with the snapshots.
+   * @return A registration object that can be used to remove the listener.
+   */
+  @NonNull
+  public ListenerRegistration addSnapshotListener(
```

Comment:
I think we also need a 

```
public ListenerRegistration addSnapshotListener(
      @NonNull LifecycleOwner lifecycleOwner, @NonNull EventListener<DocumentSnapshot> listener)
```

because there is an equivalent for `Activity`.

Reply:
Nice catch! Just added.

Reply:
Actually, one of my colleague pointed out that it might be a compilation error to try to pass `AppCompatActivity` here, because it is both `LifeCycleOwner` and `Activity`, and I just confirmed it.

I don't see an easy way out unfortunately..

Reply:
@wu-hui I forgot `AppCompatActivity` is both... I think the compilation error might occur in all of the APIs added here. Developers will need to explicitly cast it (eg. `addSnapshotListener((LifecyleOwner) appCompatActivity, listener)`)

A possible way out would be exposing one more method that takes `AppCompatActivity` as an argument and internally we would cast it to LifecycleOwner. But the problem with this is that it'll require the Firestore SDK to depend on `androidx.appcompat`.

Reply:
Right. I think neither will work perfectly.

To accommodate `LifeCycleOwner`, we might have to consider revamping the API around life cycle management for listener. I will try to make a proposal internally first. 

This does mean I cannot merge this PR as-is. Will check back with you when we have a clear idea.